igvc_navigation/src/trajectory_planner/smooth_control.cpp

```diff
@@ -0,0 +1,211 @@
+#include "smooth_control.h"
+#include <igvc_utils/robot_state.h>
+#include <algorithm>
+
+SmoothControl::SmoothControl(SmoothControlOptions smooth_control_options, PathGenerationOptions path_generation_options,
+                             TargetSelectionOptions target_selection_options,
+                             CurvatureBlendingOptions curvature_blending_options, double axle_length)
+  : smooth_control_options_{ smooth_control_options }
+  , path_generation_options_{ path_generation_options }
+  , target_selection_options_{ target_selection_options }
+  , curvature_blending_options_{ curvature_blending_options }
+  , axle_length_{ axle_length }
+{
+  ros::NodeHandle pNh{ "~" };
+  target_pub_ = pNh.advertise<nav_msgs::Path>("smooth_control/target", 1);
+  closest_point_pub_ = pNh.advertise<nav_msgs::Path>("smooth_control/closest_point", 1);
+}
+
+void SmoothControl::getPath(const nav_msgs::PathConstPtr& path, const igvc_msgs::trajectoryPtr& trajectory_ptr,
+                            const RobotState& start_pos)
+{
+  RobotState state = start_pos;
+  std::optional<RobotState> simulation_target = target_;
+  size_t path_index = 0;
+
+  ros::Time time = path->header.stamp;
+  trajectory_ptr->header.stamp = path->header.stamp;
+
+  // Calculate timesteps
+  static ros::Duration dt = ros::Duration(1 / path_generation_options_.simulation_frequency);
+
+  // Visualization
+  nav_msgs::Path closest_point_path;
+  nav_msgs::Path targets_path;
+
+  closest_point_path.header = path->header;
+  targets_path.header = path->header;
+  for (int i = 0; i < path_generation_options_.getNumSamples() && path_index < path->poses.size() - 1; i++, time += dt)
```

Comment:
```cpp
// Populate the trajectory with samples generated by the smooth control law
```
Also, it makes a lot more sense to calculate time in the loop with `ros::Time time = path->header.stamp + i * dt` instead of maintaining two separate iteration variables.

Reply:
yeah, I'll move `time` into the loop body to make it clearer.